cpp/src/Slice/Grammar.y

```diff
@@ -1642,6 +1661,57 @@ exception
 }
 ;
 
+// ----------------------------------------------------------------------
+type_alias_def
+// ----------------------------------------------------------------------
+: ICE_USING ICE_IDENTIFIER '=' local_metadata type
+{
+    StringTokPtr ident = StringTokPtr::dynamicCast($2);
+    StringListTokPtr metadata = StringListTokPtr::dynamicCast($4);
+
+    if (TypePtr type = TypePtr::dynamicCast($5))
+    {
+        if (auto alias = TypeAliasPtr::dynamicCast(type))
+        {
+            mergeMetadataInPlace(metadata->v, alias->typeMetadata());
+            type = alias->underlying();
+        }
+
+        // Metadata cannot be specified on aliases of classes and interfaces.
+        if (!metadata->v.empty())
+        {
+            if (ClassDeclPtr::dynamicCast(type))
+            {
+                unit->error("illegal metadata on `" + ident->v +
+                            "': metadata cannot be specified on aliases of classes");
+            }
+            else if (InterfaceDeclPtr::dynamicCast(type))
+            {
+                unit->error("illegal metadata on `" + ident->v +
+                            "': metadata cannot be specified on aliases of interfaces");
+            }
+        }
+
+        ModulePtr cont = unit->currentModule();
+        $$ = cont->createTypeAlias(ident->v, type, metadata->v);
+    }
+    else
+    {
+        $$ = nullptr;
```

Comment:
Isn't this an error?

Reply:
Yes, but we report errors like this when the type resolution fails, so by the time we get here, we've already issued one.

However, we still need to explicitly set `$$ = nullptr` here, because if Bison reaches the end of a statement and `$$` hasn't been set, it will implicitly set it to `$$ = $1`, which would be the `ICE_USING` token here, and which we don't want to pass up the parsing tree.